File: providers/aws/src/test/java/cloudspec/aws/ec2/EC2NetworkInterfaceLoaderTest.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2;

import cloudspec.aws.ec2.loader.EC2NetworkInterfaceLoader;
import datagen.services.ec2.model.NetworkInterfaceGenerator;
import org.junit.Test;
import software.amazon.awssdk.services.ec2.model.DescribeNetworkInterfacesRequest;
import software.amazon.awssdk.services.ec2.model.DescribeNetworkInterfacesResponse;
import software.amazon.awssdk.services.ec2.model.NetworkInterface;

import java.util.function.Consumer;

import static org.junit.Assert.*;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.doAnswer;

public class EC2NetworkInterfaceLoaderTest extends EC2LoaderTest {
    private final EC2NetworkInterfaceLoader loader = new EC2NetworkInterfaceLoader(clientsProvider);

    @Test
    public void shouldLoadNetworkInterfaces() {
        var networkInterfaces = NetworkInterfaceGenerator.networkInterfaces();
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeNetworkInterfacesRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeNetworkInterfacesRequest.builder();
            builderLambda.accept(builder);
            assertTrue(builder.build().filters().isEmpty());
            return DescribeNetworkInterfacesResponse.builder()
                                                    .networkInterfaces(networkInterfaces.toArray(new NetworkInterface[0]))
                                                    .build();
        }).when(ec2Client).describeNetworkInterfaces(any(Consumer.class));

        var resources = loader.getAll();

        assertNotNull(resources);
        assertEquals(networkInterfaces.size(), resources.size());
        assertTrue(networkInterfaces.containsAll(resources));
    }

    @Test
    public void shouldNotLoadNetworkInterfaceByRandomId() {
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeNetworkInterfacesRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeNetworkInterfacesRequest.builder();
            builderLambda.accept(builder);
            assertFalse(builder.build().filters().isEmpty());
            return DescribeNetworkInterfacesResponse.builder()
                                                    .build();
        }).when(ec2Client).describeNetworkInterfaces(any(Consumer.class));

        var resourceOpt = loader.getById(NetworkInterfaceGenerator.networkInterfaceId());

        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isEmpty());
    }

    @Test
    public void shouldLoadNetworkInterfaceById() {
        var networkInterfaces = NetworkInterfaceGenerator.networkInterfaces(1);
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeNetworkInterfacesRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeNetworkInterfacesRequest.builder();
            builderLambda.accept(builder);
            assertFalse(builder.build().filters().isEmpty());
            return DescribeNetworkInterfacesResponse.builder()
                                                    .networkInterfaces(networkInterfaces.toArray(new NetworkInterface[0]))
                                                    .build();
        }).when(ec2Client).describeNetworkInterfaces(any(Consumer.class));

        var networkInterface = networkInterfaces.get(0);
        var resourceOpt = loader.getById(networkInterface.networkInterfaceId());
        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isPresent());
        assertEquals(resourceOpt.get(), networkInterface);
    }
}

File: providers/aws/src/test/kotlin/cloudspec/aws/ec2/EC2NetworkInterfaceLoaderTest.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2

import arrow.core.None
import arrow.core.Some
import datagen.services.ec2.model.NetworkInterfaceGenerator.networkInterfaceId
import datagen.services.ec2.model.NetworkInterfaceGenerator.networkInterfaces
import io.mockk.every
import org.junit.Test
import software.amazon.awssdk.services.ec2.model.DescribeNetworkInterfacesRequest
import software.amazon.awssdk.services.ec2.model.DescribeNetworkInterfacesResponse
import java.util.function.Consumer
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class EC2NetworkInterfaceLoaderTest : EC2LoaderTest() {
    private val loader = EC2NetworkInterfaceLoader(clientsProvider)

    @Test
    fun `should load all resources`() {
        val networkInterfaces = networkInterfaces()

        every {
            ec2Client.describeNetworkInterfaces(any<Consumer<DescribeNetworkInterfacesRequest.Builder>>())
        } answers {
            val builder = DescribeNetworkInterfacesRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeNetworkInterfacesRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertTrue(request.filters().isEmpty())

            DescribeNetworkInterfacesResponse.builder()
                    .networkInterfaces(networkInterfaces)
                    .build()
        }

        val resources = loader.all
        assertNotNull(resources)
        assertEquals(networkInterfaces.size, resources.size)
    }

    @Test
    fun `should not load resource by random id`() {
        every {
            ec2Client.describeNetworkInterfaces(any<Consumer<DescribeNetworkInterfacesRequest.Builder>>())
        } answers {
            val builder = DescribeNetworkInterfacesRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeNetworkInterfacesRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeNetworkInterfacesResponse.builder()
                    .build()
        }

        val resourceOpt = loader.byId(networkInterfaceId())
        assertTrue(resourceOpt is None)
    }

    @Test
    fun `should load resource by id`() {
        val networkInterfaces = networkInterfaces()

        every {
            ec2Client.describeNetworkInterfaces(any<Consumer<DescribeNetworkInterfacesRequest.Builder>>())
        } answers {
            val builder = DescribeNetworkInterfacesRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeNetworkInterfacesRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeNetworkInterfacesResponse.builder()
                    .networkInterfaces(networkInterfaces)
                    .build()
        }

        val networkInterface = networkInterfaces[0]
        val resourceOpt = loader.byId(networkInterface.networkInterfaceId())
        assertTrue(resourceOpt is Some<*>)
    }
}
